src/goto-programs/goto_contractor.h

```diff
@@ -0,0 +1,257 @@
+//
+// Created by Mohannad Aldughaim on 09/01/2022.
+//
+
+#ifndef ESBMC_GOTO_CONTRACTOR_H
+#define ESBMC_GOTO_CONTRACTOR_H
+
+#include "goto_k_induction.h"
+#include <iostream>
+#include "util/goto_expr_factory.h"
+#include "goto_functions.h"
+#include <util/algorithms.h>
+#include <util/message/message.h>
+#include <goto-programs/goto_loops.h>
+#include <goto-programs/remove_skip.h>
+#include <goto-programs/goto_functions.h>
+#include <goto-programs/goto_functions.h>
+#include <goto-programs/loopst.h>
+#include <util/message/message_stream.h>
+#include <util/std_types.h>
+#include <ibex.h>
+#include <ibex/ibex_Interval.h>
+#include <ibex/ibex_Expr.h>
+#include <ibex/ibex_Ctc.h>
+#include "irep2/irep2.h"
+#include "util/type_byte_size.h"
+
+#define MAX_VAR 10
+#define NOT_FOUND -1
+
+using namespace ibex;
+
+/**
+ * This class is for mapping the variables with their names and intervals.
+ * It includes functionalities such as search for a variable by name and add
+ * a new variable. Also update a variable interval by changing the upper/lower
+ * limit.
+ */
+class CspMap
+{
+public:
+  ibex::IntervalVector intervals;
+  std::string var_name[MAX_VAR];
+  symbol2tc symbols[MAX_VAR];
+  bool nondecreasing[MAX_VAR], nonincreasing[MAX_VAR];
+
+  CspMap()
+  {
+    intervals.resize(MAX_VAR);
+    for(int i=0;i<MAX_VAR;i++)
+    nondecreasing[i] = nonincreasing[i] = false;
+  }
+  int add_var(std::string name, symbol2t symbol)
+  {
+    if(find(name) == NOT_FOUND && n < MAX_VAR)
+    {
+      symbols[n] = symbol;
+      var_name[n] = name;
+
+      ///keep for later. set initial intervals
+      //auto w = symbol.type->get_width();
+      /*if(is_signedbv_type(symbol.type))
+        add_interval(-pow(2,w-1),pow(2,w-1)-1,n);
+      else
+        add_interval(0,pow(2,w)-1,n);*/
+      n++;
+      return n - 1;
+    }
+    return NOT_FOUND;
+  }
+  void add_interval(double lb, double ub, int index)
+  {
+    intervals[index] = Interval(lb, ub);
+  }
+  void update_lb_interval(double lb, int index)
+  {
+    add_interval(lb, intervals[index].ub(), index);
+  }
+  void update_ub_interval(double ub, int index)
+  {
+    add_interval(intervals[index].lb(), ub, index);
+  }
+  int find(std::string name)
+  {
+    for(size_t i = 0; i < n; i++)
+      if(var_name[i] == name)
```

Comment:
Isn't it possible to use a proper `std::map` or even `std::unordered_map` instead of this linear search using string comparisons? That way you might also be able to get rid of the rather arbitrary `MAX_VAR` limit.

Reply:
done